File: services/tradelog/tradelog-api/src/main/java/com/example/tradelog/api/facade/JournalFacade.java
package com.example.tradelog.api.facade;

import com.example.tradelog.api.service.DividendJournalService;
import com.example.tradelog.api.service.OptionJournalService;
import com.example.tradelog.api.service.ShareJournalService;
import com.example.tradelog.api.service.TransactionJournalService;
import com.example.tradelog.api.spec.model.DividendJournalModel;
import com.example.tradelog.api.spec.model.OptionJournalModel;
import com.example.tradelog.api.spec.model.ShareJournalModel;
import com.example.tradelog.api.spec.model.TradeSummaryModel;
import com.example.tradelog.api.spec.model.TransactionSettingsModel;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class JournalFacade {

    private TransactionJournalService transactionService;
    private ShareJournalService shareService;
    private OptionJournalService optionService;
    private DividendJournalService dividendService;

    public JournalFacade(TransactionJournalService transactionService, ShareJournalService shareService, OptionJournalService optionService, DividendJournalService dividendService) {
        this.transactionService = transactionService;
        this.shareService = shareService;
        this.optionService = optionService;
        this.dividendService = dividendService;
    }


    public List<String> getAllTradedSymbols(String accountId) {
        return transactionService.getAllTradedSymbols(accountId);
    }

    public List<TradeSummaryModel> getSummary(String accountId) {
        Map<String, TradeSummaryModel> shareSummaries = this.shareService.getSummaries(accountId);
        Map<String, TradeSummaryModel> optionSummaries = this.optionService.getSummaries(accountId);
        Map<String, TradeSummaryModel> dividendSummaries = this.dividendService.getSummaries(accountId);

        Map<String, TradeSummaryModel> summaryModelMap = new HashMap<>();
        shareSummaries.forEach(summaryModelMap::put);

        optionSummaries.forEach( (k, v) -> {
            if (summaryModelMap.containsKey(v.getSymbol())) {
                TradeSummaryModel storedModel = summaryModelMap.get(v.getSymbol());
                summaryModelMap.put(v.getSymbol(), new TradeSummaryModel(storedModel.getSymbol(), storedModel.getTrades() + v.getTrades(), storedModel.getTotal().add(v.getTotal())));
            } else {
                summaryModelMap.put(v.getSymbol(), v);
            }
        });

        dividendSummaries.forEach( (k, v) -> {
            if (summaryModelMap.containsKey(v.getSymbol())) {
                TradeSummaryModel storedModel = summaryModelMap.get(v.getSymbol());
                summaryModelMap.put(v.getSymbol(), new TradeSummaryModel(storedModel.getSymbol(), storedModel.getTrades() + v.getTrades(), storedModel.getTotal().add(v.getTotal())));
            } else {
                summaryModelMap.put(v.getSymbol(), v);
            }
        });


        return summaryModelMap.entrySet().stream()
                .sorted(Map.Entry.comparingByKey())
                .map(Map.Entry::getValue)
                .collect(Collectors.toList());
    }

    public boolean updateOptions(TransactionSettingsModel model) {
        return transactionService.updateOptions(model);
    }

    public void updateSettingsBulk(List<TransactionSettingsModel> models) {
        transactionService.updateSettingsBulk(models);
    }

    public List<DividendJournalModel> getAllDividendsBySymbol(String accountId, String symbol) {
        return dividendService.getAllBySymbol(accountId, symbol);
    }

    public Optional<DividendJournalModel> createDividendRecord(DividendJournalModel model) {
        Optional<String> optionalId = transactionService.createTransactionRecord(model.getTransactionDetails());
        if (optionalId.isPresent()) {
            return dividendService.createRecord(optionalId.get(), model);
        }
        return Optional.empty();
    }

    //TODO: transactional
    public boolean deleteDividendRecord(String accountId, String transactionId) {
        return dividendService.deleteRecord(transactionId)
                && transactionService.deleteSettingsRecord(transactionId)
                && transactionService.deleteDividendRecord(transactionId, accountId);
    }


    public List<ShareJournalModel> getAllSharesBySymbol(String accountId, String symbol) {
        return shareService.getAllBySymbol(accountId, symbol);
    }

    public Optional<ShareJournalModel> createShareRecord(ShareJournalModel model) {
        Optional<String> optionalId = transactionService.createTransactionRecord(model.getTransactionDetails());
        if (optionalId.isPresent()) {
            return shareService.createRecord(optionalId.get(), model);
        }
        return Optional.empty();
    }

    //TODO: should be transactional
    public boolean deleteShareRecord(String accountId, String transactionId) {
        return shareService.deleteRecord(transactionId)
                && transactionService.deleteSettingsRecord(transactionId)
                && transactionService.deleteShareRecord(transactionId, accountId);
    }

    public List<OptionJournalModel> getAllOptionsBySymbol(String accountId, String symbol) {
        return optionService.getAllBySymbol(accountId, symbol);
    }

    public Optional<OptionJournalModel> createOptionRecord(OptionJournalModel model) {
        Optional<String> optionalId = transactionService.createTransactionRecord(model.getTransactionDetails());
        if (optionalId.isPresent()) {
            return optionService.createRecord(optionalId.get(), model);
        }
        return Optional.empty();
    }

    public boolean deleteOptionRecord(String accountId, String transactionId) {
        return optionService.deleteRecord(transactionId)
                && transactionService.deleteSettingsRecord(transactionId)
                && transactionService.deleteOptionRecord(transactionId, accountId);
    }
}

File: services/tradelog/tradelog-api/src/main/kotlin/com/example/tradelog/api/core/facade/JournalFacade.kt
package com.example.tradelog.api.core.facade

import arrow.core.Either
import arrow.core.flatMap
import com.example.common.service.ServiceError
import com.example.tradelog.api.core.model.*
import com.example.tradelog.api.core.service.DividendJournalService
import com.example.tradelog.api.core.service.OptionJournalService
import com.example.tradelog.api.core.service.ShareJournalService
import com.example.tradelog.api.core.service.TransactionService
import org.springframework.stereotype.Service
import java.util.*
import java.util.stream.Collectors
import kotlin.collections.HashMap

@Service
class JournalFacade(private val transactionService: TransactionService,
                    private val shareService: ShareJournalService,
                    private val optionService: OptionJournalService,
                    private val dividendService: DividendJournalService
) {
    enum class SummaryMatrixType {
        SHARES,
        OPTIONS_AND_DIVIDENDS
    }

    fun getSummary(accountId: UUID): Either<ServiceError, List<TradeSummaryModel>> {
        //TODO: do parallel call
        //TODO: rewrite it
        val shareSummaries = shareService.getSummaries(accountId).orNull() ?: return Either.Left(ServiceError.DataAccessError("Cannot get share summaries"))
        val optionSummaries = optionService.getSummaries(accountId).orNull() ?: return Either.Left(ServiceError.DataAccessError("Cannot get option summaries"))
        val dividendSummaries = dividendService.getSummaries(accountId).orNull() ?: return Either.Left(ServiceError.DataAccessError("Cannot get dividend summaries"))

        val summaryModelMap: HashMap<String, TradeSummaryModel> = HashMap()

        shareSummaries.forEach { summaryModelMap[it.key] = it.value }

        optionSummaries.forEach {
            if (summaryModelMap.containsKey(it.key)) {
                val storedModel: TradeSummaryModel = summaryModelMap[it.key]!!

                summaryModelMap[it.key] = TradeSummaryModel(symbol = it.key,
                        trades = storedModel.trades + it.value.trades,
                        total = storedModel.total.add(it.value.total),
                        legClosed = storedModel.legClosed)
            } else {
                if (it.value.legClosed) {
                    summaryModelMap[it.key] = it.value
                }
            }
        }

        dividendSummaries.forEach {
            if (summaryModelMap.containsKey(it.key)) {
                val storedModel: TradeSummaryModel = summaryModelMap[it.key]!!

                summaryModelMap[it.key] = TradeSummaryModel(symbol = it.key,
                        trades = storedModel.trades + it.value.trades,
                        total = storedModel.total.add(it.value.total),
                        legClosed = storedModel.legClosed)
            } else {
                if (it.value.legClosed) {
                    summaryModelMap[it.key] = it.value
                }
            }
        }

        return Either.Right(
                summaryModelMap.entries.stream()
                        .sorted(compareBy { it.key })
                        .map { it.value }
                        .collect(Collectors.toList())
        )
    }


    fun getSummaryMatrix(accountId: UUID, portfolioId: UUID, matrixType: SummaryMatrixType): Either<ServiceError, List<SummaryMatrixModel>> {
        val summaryMatrixList = when(matrixType) {
            SummaryMatrixType.OPTIONS_AND_DIVIDENDS -> transactionService.getOptionAndDividendSummaryMatrix(accountId, portfolioId)
            SummaryMatrixType.SHARES -> transactionService.getSharesSummaryMatrix(accountId, portfolioId)
        }


        val groupedMap: (List<SummaryMatrixModel>) -> Map<Pair<Int, Int>, Double> = {
            models -> groupByYearAndMonth(models)
        }

        return summaryMatrixList
                .map { groupedMap(it) }
                .map {
                    it.entries.map { entry -> SummaryMatrixModel(year = entry.key.first, month = entry.key.second, total = entry.value) }
                }
    }

    private fun groupByYearAndMonth(models: List<SummaryMatrixModel>): Map<Pair<Int, Int>, Double> {
        val map = HashMap<Pair<Int, Int>, Double>()
        models.forEach {
            if (map.containsKey(Pair(it.year, it.month))) {
                val storedTotal = map.getValue(Pair(it.year, it.month))
                map[Pair(it.year, it.month)] = storedTotal.plus(it.total)
            } else {
                map[Pair(it.year, it.month)] = it.total
            }
        }

        return map
    }


    fun createShareRecord(model: ShareJournalModel): Either<ServiceError, ShareJournalModel> {
        val createTransactionRecord = transactionService.createRecord(model.transactionDetails)
        val createShareRecord: (UUID, ShareJournalModel) -> Either<ServiceError, ShareJournalModel> = { id, journalModel ->
            shareService.createRecord(transactionId = id, model = journalModel)
        }

        return createTransactionRecord
                .flatMap { createShareRecord(it, model) }
    }

    fun editShareRecord(transactionId: UUID, model: ShareJournalModel): Either<ServiceError, Unit> {
        if (transactionId != model.transactionDetails.id) {
            return Either.Left(ServiceError.ValidationError("Transaction ID is not valid"))
        }

        return transactionService.editRecord(model.transactionDetails)
                .flatMap { shareService.editRecord(model) }
    }

    fun deleteShareRecord(accountId: UUID, transactionId: UUID): Either<ServiceError, Unit> {
        return shareService.getById(accountId, transactionId)
                .flatMap { shareService.deleteRecord(transactionId) }
                .flatMap { transactionService.deleteSettings(transactionId) }
                .flatMap { transactionService.deleteRecord(accountId, transactionId) }
    }


    fun createOptionRecord(model: OptionJournalModel): Either<ServiceError, OptionJournalModel> {
        val createTransactionRecord = transactionService.createRecord(model.transactionDetails)
        val createOptionRecord: (UUID, OptionJournalModel) -> Either<ServiceError, OptionJournalModel> = { id, journalModel ->
            optionService.createRecord(transactionId = id, model = journalModel)
        }

        return createTransactionRecord
                .flatMap { createOptionRecord(it, model) }
    }

    fun editOptionRecord(transactionId: UUID, model: OptionJournalModel): Either<ServiceError, Unit> {
        if (transactionId != model.transactionDetails.id) {
            return Either.Left(ServiceError.ValidationError("Transaction ID is not valid"))
        }

        return transactionService.editRecord(model.transactionDetails)
                .flatMap { optionService.editRecord(model) }
    }

    fun deleteOptionRecord(accountId: UUID, transactionId: UUID): Either<ServiceError, Unit> {
        return optionService.getById(accountId, transactionId)
                .flatMap { optionService.deleteRecord(transactionId) }
                .flatMap { transactionService.deleteSettings(transactionId) }
                .flatMap { transactionService.deleteRecord(accountId, transactionId) }
    }

    fun createDividendRecord(model: DividendJournalModel): Either<ServiceError, DividendJournalModel> {
        val createTransactionRecord = transactionService.createRecord(model.transactionDetails)
        val createDividendRecord: (UUID, DividendJournalModel) -> Either<ServiceError, DividendJournalModel> = { id, journalModel ->
            dividendService.createRecord(transactionId = id, model = journalModel)
        }

        return createTransactionRecord
                .flatMap { createDividendRecord(it, model) }
    }

    fun editDividendRecord(transactionId: UUID, model: DividendJournalModel): Either<ServiceError, Unit> {
        if (transactionId != model.transactionDetails.id) {
            return Either.Left(ServiceError.ValidationError("Transaction ID is not valid"))
        }

        return transactionService.editRecord(model.transactionDetails)
                .flatMap { dividendService.editRecord(model) }
    }

    fun deleteDividendRecord(accountId: UUID, transactionId: UUID): Either<ServiceError, Unit> {
        return dividendService.getById(accountId, transactionId)
                .flatMap { dividendService.deleteRecord(transactionId) }
                .flatMap { transactionService.deleteSettings(transactionId) }
                .flatMap { transactionService.deleteRecord(accountId, transactionId) }
    }
}
